python-avd/pyavd/api/pool_manager/base_classes.py

```diff
@@ -0,0 +1,282 @@
+        return False
+
+    def _next_available(self) -> T_ValueType:
+        if self.collection.value_type is int:
+            collection = cast(PoolCollection[T_PoolKeyType, T_AssignmentKeyType, int], self.collection)
+            assignments = cast(dict[T_AssignmentKeyType, PoolAssignment[T_AssignmentKeyType, int]], self.assignments)
+            existing_ids = {assignment.value for assignment in assignments.values()}
+            available_ids = set(range(collection.min_value, len(existing_ids) + 2)) - existing_ids
+            next_available = next(iter(available_ids))
+            if collection.max_value is not None and next_available > collection.max_value:
+                msg = (
+                    f"No available values found within '{self.collection.pools_key}' with key '{self.pool_key}' "
+                    f"and range {collection.min_value}-{collection.max_value}."
+                )
+                raise ValueError(msg)
+            return cast(T_ValueType, next_available)
+
+        msg = f"Currently we only support pools of Integers. Got {self.collection.value_type}."
+        raise NotImplementedError(msg)
+
+    def get_assignment(self, key: T_AssignmentKeyType, requested_value: T_ValueType | None = None) -> PoolAssignment[T_AssignmentKeyType, T_ValueType]:
+        """
+        Returns the assignment for the key if found in the pool.
+
+        Otherwise a new entry is inserted into the pool and returned.
+
+        If 'requested_value' is given and available in the pool, any existing assignment will be updated and returned.
+        For new assignments this value will be used.
+        """
+        if key not in self.assignments:
+            # No existing assignment. Use the requested value if given and available. Otherwise use next available.
+            new_assignment_value = requested_value if requested_value is not None and self._is_value_available(requested_value) else self._next_available()
+            self.assignments[key] = self.collection.assignment_cls(key=key, value=new_assignment_value, accessed=True)
+            self.collection.changed = True
+        elif requested_value is not None and requested_value != self.assignments[key].value and self._is_value_available(requested_value):
+            # Existing assignment but not with the requested value. Change the existing assignment to the requested_value.
+            self.assignments[key] = self.collection.assignment_cls(key=key, value=requested_value, accessed=True)
+            self.collection.changed = True
+        else:
+            # Existing item
+            self.assignments[key].accessed = True
+
+        return self.assignments[key]
+
+    def _remove_stale_assignments(self) -> None:
+        """
+        Remove stale assignments. An assignment is deemed stale if "accessed" is not True. Sets parent .changed if it removed anything.
+
+        Note: This method is called from the parent collection and should _not_ be called manually.
```

Comment:
should it be moved to the parent collection class then? as it is really the object which has this responsability

Reply:
Yeah I guess it could now when we track everything from the collection layer.

Reply:
these are nested through separate layers, and it is up to the resource (could be subclassed) to clean up their assignments.

Reply:
ack